File: go/cmd/minimax/commands/text.go
package commands

import (
	"context"
	"fmt"
	"time"

	"github.com/spf13/cobra"

	"github.com/haivivi/giztoy/pkg/minimax"
)

var textCmd = &cobra.Command{
	Use:   "text",
	Short: "Text generation service",
	Long: `Text generation service (chat completions).

Supports streaming and non-streaming text generation with tool calling.

Example request file (chat.yaml):
  model: MiniMax-M2.1
  messages:
    - role: system
      content: You are a helpful assistant.
    - role: user
      content: Hello, who are you?
  max_tokens: 1000
  temperature: 0.7`,
}

var textChatCmd = &cobra.Command{
	Use:   "chat",
	Short: "Create a chat completion",
	Long: `Create a chat completion.

Examples:
  minimax -c myctx text chat -f chat.yaml
  minimax -c myctx text chat -f chat.yaml --json | jq '.choices[0].message'`,
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := requireInputFile(); err != nil {
			return err
		}

		ctx, err := getContext()
		if err != nil {
			return err
		}

		var req minimax.ChatCompletionRequest
		if err := loadRequest(getInputFile(), &req); err != nil {
			return err
		}

		// Use default model if not specified
		if req.Model == "" {
			if defaultModel := ctx.GetExtra("default_model"); defaultModel != "" {
				req.Model = defaultModel
			} else {
				req.Model = minimax.ModelM2_1
			}
		}

		printVerbose("Using context: %s", ctx.Name)
		printVerbose("Model: %s", req.Model)
		printVerbose("Messages: %d", len(req.Messages))

		// Create API client
		client := createClient(ctx)

		// Call API
		reqCtx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
		defer cancel()

		resp, err := client.Text.CreateChatCompletion(reqCtx, &req)
		if err != nil {
			return fmt.Errorf("chat completion failed: %w", err)
		}

		return outputResult(resp, getOutputFile(), isJSONOutput())
	},
}

var textChatStreamCmd = &cobra.Command{
	Use:   "chat-stream",
	Short: "Create a streaming chat completion",
	Long: `Create a streaming chat completion.

The response will be streamed to stdout in real-time.

Examples:
  minimax -c myctx text chat-stream -f chat.yaml`,
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := requireInputFile(); err != nil {
			return err
		}

		ctx, err := getContext()
		if err != nil {
			return err
		}

		var req minimax.ChatCompletionRequest
		if err := loadRequest(getInputFile(), &req); err != nil {
			return err
		}

		// Use default model if not specified
		if req.Model == "" {
			if defaultModel := ctx.GetExtra("default_model"); defaultModel != "" {
				req.Model = defaultModel
			} else {
				req.Model = minimax.ModelM2_1
			}
		}

		printVerbose("Using context: %s", ctx.Name)
		printVerbose("Model: %s", req.Model)
		printVerbose("Streaming mode enabled")

		// Create API client
		client := createClient(ctx)

		// Call streaming API
		reqCtx, cancel := context.WithTimeout(context.Background(), 300*time.Second)
		defer cancel()

		var fullContent string
		for chunk, err := range client.Text.CreateChatCompletionStream(reqCtx, &req) {
			if err != nil {
				return fmt.Errorf("streaming failed: %w", err)
			}
			if len(chunk.Choices) > 0 && chunk.Choices[0].Delta != nil {
				content := chunk.Choices[0].Delta.Content
				fmt.Print(content)
				fullContent += content
			}
		}
		fmt.Println() // New line after streaming

		printVerbose("Total content length: %d characters", len(fullContent))

		return nil
	},
}

func init() {
	textCmd.AddCommand(textChatCmd)
	textCmd.AddCommand(textChatStreamCmd)
}

File: rust/cmd/minimax/src/commands/text.rs
//! Text generation commands.
//!
//! Compatible with Go version's text commands.

use std::io::Write;
use std::pin::pin;

use clap::{Args, Subcommand};
use futures::StreamExt;

use giztoy_minimax::{ChatCompletionRequest, MODEL_M2_1};

use super::{
    create_client, get_context, load_request, output_result, print_verbose, require_input_file,
};
use crate::Cli;

/// Text generation service.
///
/// Supports chat completions with streaming.
#[derive(Args)]
pub struct TextCommand {
    #[command(subcommand)]
    command: TextSubcommand,
}

#[derive(Subcommand)]
enum TextSubcommand {
    /// Create a chat completion
    Chat,
    /// Create a streaming chat completion
    Stream,
}

impl TextCommand {
    pub async fn run(&self, cli: &Cli) -> anyhow::Result<()> {
        match &self.command {
            TextSubcommand::Chat => self.chat(cli).await,
            TextSubcommand::Stream => self.stream(cli).await,
        }
    }

    async fn chat(&self, cli: &Cli) -> anyhow::Result<()> {
        let input_file = require_input_file(cli)?;
        let ctx = get_context(cli)?;

        let mut req: ChatCompletionRequest = load_request(input_file)?;

        // Use defaults if not specified
        if req.model.is_empty() {
            req.model = MODEL_M2_1.to_string();
        }

        print_verbose(cli, &format!("Using context: {}", ctx.name));
        print_verbose(cli, &format!("Model: {}", req.model));
        print_verbose(cli, &format!("Messages: {}", req.messages.len()));

        let client = create_client(&ctx)?;
        let resp = client.text().create_chat_completion(&req).await?;

        // Output result
        output_result(&resp, cli.output.as_deref(), cli.json)
    }

    async fn stream(&self, cli: &Cli) -> anyhow::Result<()> {
        let input_file = require_input_file(cli)?;
        let ctx = get_context(cli)?;

        let mut req: ChatCompletionRequest = load_request(input_file)?;

        // Use defaults if not specified
        if req.model.is_empty() {
            req.model = MODEL_M2_1.to_string();
        }

        print_verbose(cli, &format!("Using context: {}", ctx.name));
        print_verbose(cli, &format!("Model: {}", req.model));

        let client = create_client(&ctx)?;

        // Streaming output
        if cli.json {
            // JSON mode: output each chunk as JSON
            let stream = client.text().create_chat_completion_stream(&req).await?;
            let mut stream = pin!(stream);

            while let Some(chunk) = stream.next().await {
                let chunk = chunk?;
                let json = serde_json::to_string(&chunk)?;
                println!("{}", json);
            }
        } else {
            // Text mode: output content directly
            let stream = client.text().create_chat_completion_stream(&req).await?;
            let mut stream = pin!(stream);

            while let Some(chunk) = stream.next().await {
                let chunk = chunk?;
                if let Some(choice) = chunk.choices.first() {
                    if let Some(content) = &choice.delta.content {
                        print!("{}", content);
                        std::io::stdout().flush()?;
                    }
                }
            }
            println!();
        }

        Ok(())
    }
}
